packages/react-router-dom/__tests__/special-characters-test.tsx

```diff
@@ -0,0 +1,712 @@
+/* eslint-disable jest/expect-expect */
+
+import { JSDOM } from "jsdom";
+import * as React from "react";
+import {
+  cleanup,
+  render,
+  fireEvent,
+  waitFor,
+  screen,
+} from "@testing-library/react";
+import type { Location, Params } from "react-router-dom";
+import {
+  BrowserRouter,
+  Link,
+  Routes,
+  Route,
+  RouterProvider,
+  createBrowserRouter,
+  createRoutesFromElements,
+  useLocation,
+  useParams,
+} from "react-router-dom";
+
+/**
+ * Here's all the special characters we want to test against.  This list was
+ * generated using the following utility in the Chrome DevTools console for
+ * maximum accuracy.  This is instead of programmatically generating during
+ * these tests where JSDOM or a bad URL polyfill might not be trustworthy.
+ *
+ * function generateCharDef(char) {
+ *   return {
+ *       char,
+ *       pathChar: new URL('/' + char, window.location.origin).pathname.replace(/^\//, ''),
+ *       searchChar: new URL('/?=' + char, window.location.origin).search.replace(/^\?=/, ''),
+ *       hashChar: new URL('/#' + char, window.location.origin).hash.replace(/^#/, ''),
+ *   };
+ * }
+ */
+
+let specialChars = [
```

Comment:
Large set of special character "definitions" of the character, and what we expect it to show up as in `location.pathname`, `location.search`, and `location.hash` since it varies according to the spec: https://url.spec.whatwg.org/#c0-control-percent-encode-set